waddrmgr/db.go

```diff
@@ -809,6 +826,78 @@ func serializeDefaultAccountRow(encryptedPubKey, encryptedPrivKey []byte,
 	return rawData
 }
 
+// deserializeWatchOnlyAccountRow deserializes the raw data from the passed
+// account row as a watch-only account.
+func deserializeWatchOnlyAccountRow(accountID []byte,
+	row *dbAccountRow) (*dbWatchOnlyAccountRow, error) {
+
+	// The serialized BIP0044 watch-only account raw data format is:
+	//   <encpubkeylen><encpubkey><masterkeyfingerprint><nextextidx>
+	//   <nextintidx><namelen><name>
+	//
+	// 4 bytes encrypted pubkey len + encrypted pubkey + 4 bytes master key
+	// fingerprint + 4 bytes next external index + 4 bytes next internal
+	// index + 4 bytes name len + name
+
+	// Given the above, the length of the entry must be at a minimum
+	// the constant value sizes.
+	if len(row.rawData) < 20 {
+		str := fmt.Sprintf("malformed serialized watch-only account "+
+			"for key %x", accountID)
+		return nil, managerError(ErrDatabase, str, nil)
+	}
+
+	retRow := dbWatchOnlyAccountRow{
+		dbAccountRow: *row,
+	}
+
+	pubLen := binary.LittleEndian.Uint32(row.rawData[0:4])
+	retRow.pubKeyEncrypted = make([]byte, pubLen)
+	copy(retRow.pubKeyEncrypted, row.rawData[4:4+pubLen])
+	offset := 4 + pubLen
+	retRow.masterKeyFingerprint = binary.LittleEndian.Uint32(row.rawData[offset : offset+4])
+	offset += 4
+	retRow.nextExternalIndex = binary.LittleEndian.Uint32(row.rawData[offset : offset+4])
+	offset += 4
+	retRow.nextInternalIndex = binary.LittleEndian.Uint32(row.rawData[offset : offset+4])
+	offset += 4
+	nameLen := binary.LittleEndian.Uint32(row.rawData[offset : offset+4])
+	offset += 4
+	retRow.name = string(row.rawData[offset : offset+nameLen])
+
+	return &retRow, nil
+}
+
+// serializeWatchOnlyAccountRow returns the serialization of the raw data field
+// for a watch-only account.
+func serializeWatchOnlyAccountRow(encryptedPubKey []byte, masterKeyFingerprint,
+	nextExternalIndex, nextInternalIndex uint32, name string) []byte {
+
+	// The serialized BIP0044 account raw data format is:
+	//   <encpubkeylen><encpubkey><masterkeyfingerprint><nextextidx>
+	//   <nextintidx><namelen><name>
+	//
+	// 4 bytes encrypted pubkey len + encrypted pubkey + 4 bytes master key
+	// fingerprint + 4 bytes next external index + 4 bytes next internal
+	// index + 4 bytes name len + name
+	pubLen := uint32(len(encryptedPubKey))
+	nameLen := uint32(len(name))
+	rawData := make([]byte, 20+pubLen+nameLen)
+	binary.LittleEndian.PutUint32(rawData[0:4], pubLen)
+	copy(rawData[4:4+pubLen], encryptedPubKey)
+	offset := 4 + pubLen
+	binary.LittleEndian.PutUint32(rawData[offset:offset+4], masterKeyFingerprint)
```

Comment:
Occurs to me that we can make all this a bit more readable by instead using `binary.Read/Write`, but we can save that for a future refactoring.

Reply:
Done.